examples/direct_fidelity_estimation.py

```diff
@@ -29,17 +29,16 @@ def compute_characteristic_function(circuit, P_i, qubits, noise):
   n = len(P_i)
   d = 2**n
 
-  trace = 0  # rho_i / sigma_i in https://arxiv.org/pdf/1104.3835.pdf
-  for x in range(d):
-    xbin = numpy.binary_repr(x, width=n)
+  simulator = cirq.DensityMatrixSimulator()
+  density_matrix = 0  #  rho or sigma in https://arxiv.org/pdf/1104.3835.pdf
+  for step in simulator.simulate_moment_steps(circuit):
+    density_matrix = step.density_matrix()
 
-    pauli_string = cirq.PauliString(dict(zip(qubits, P_i)))
-    display = cirq.approx_pauli_string_expectation(pauli_string, num_samples=1)
-    y = cirq.sample(cirq.Circuit(
-        pauli_string,
-        circuit)).measurements['y'][0]
-
-    trace += sum([int(xbin[i]) == y[i] for i in range(n)])
+  pauli_string = cirq.PauliString(dict(zip(qubits, P_i)))
```

Comment:
add type annotations to this function's signature. They help a lot. In this case, when reading the code I'm wondering "what is `P_i` and why isn't it a `PauliString` already"

Reply:
Done.